### PR TITLE
Upgraded Practitioner/Provider Resource

### PR DESCRIPTION
**Pull request by yashrajbothra**
Fixes #

#### Changes proposed in this pull request:

-
-
-

### CODE REVIEW

**Comment by yashrajbothra** [comment]
Still need to add some more Test Cases in [practitioners.json](https://github.com/openemr/openemr/blob/db445ca08c16c0d09beb9d165f76d06822156f4b/tests/Tests/Fixtures/practitioners.json)

File: sql/5_0_2-to-6_0_0_upgrade.sql [hunk 1]
```diff
@@ -719,4 +719,14 @@ ALTER TABLE `users` ADD `portal_user` TINYINT(1) NOT NULL DEFAULT '0';
 
 #IfMissingColumn users supervisor_id
 ALTER TABLE `users` ADD `supervisor_id` INT(11) NOT NULL DEFAULT '0';
+
+#IfMissingColumn users uuid
+ALTER TABLE `users` ADD `uuid` binary(16) DEFAULT NULL;
+#EndIf
+
+#IfUuidNeedUpdate form_encounter
```
Comment (bradymiller) [inline sql/5_0_2-to-6_0_0_upgrade.sql:727]:
```
#IfUuidNeedUpdate users
```
:)

File: sql/database.sql [hunk 2]
```diff
@@ -6698,6 +6698,7 @@ CREATE TABLE `transactions` (
 DROP TABLE IF EXISTS `users`;
 CREATE TABLE `users` (
   `id` bigint(20) NOT NULL auto_increment,
+  `uuid` binary(16) DEFAULT NULL,
```
Comment (bradymiller) [inline sql/database.sql:6701]: also needs to be set up as unique key

File: src/Services/PractitionerService.php [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            $data,
+            PractitionerValidator::DATABASE_INSERT_CONTEXT
+        );
+
+        if (!$processingResult->isValid()) {
+            return $processingResult;
+        }
+
+        $freshPid = $this->getFreshId("id", "users");
+        $data['id'] = $freshPid;
```
Comment (bradymiller) [inline src/Services/PractitionerService.php:200]: Note that id in users table is autoincrementing, so no reason to create an id here (doing this opens door to possible concurrency issues). Can just let database create the id naturally on the insert.

Comment (yashrajbothra) [reply src/Services/PractitionerService.php:200]: Agreed :+1:  

Thanks :)

File: src/Services/PractitionerService.php [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        $sql .= $query['set'];
+
+        $results = sqlInsert(
+            $sql,
+            $query['bind']
+        );
+
+        if ($results) {
+            $processingResult->addData(array(
+                'id' => $freshPid,
```
Comment (bradymiller) [inline src/Services/PractitionerService.php:214]: the $results should be storing the id (this is the return from the sqlInsert call).

Comment (bradymiller) [reply src/Services/PractitionerService.php:214]: so this should work:
```
'id' => $results,
```

File: tests/Tests/Api/PractitionerApiTest.php [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+namespace OpenEMR\Tests\Api;
+
+use PHPUnit\Framework\TestCase;
+use OpenEMR\Tests\Api\ApiTestClient;
+use OpenEMR\Tests\Fixtures\PractitionerFixtureManager;
+
+/**
+ * Practitioner API Endpoint Test Cases.
+ * NOTE: currently disabled (by naming convention) until work is completed to support running as part of Travis CI
+ * @coversDefaultClass OpenEMR\Tests\Api\ApiTestClient
```
Comment (bradymiller) [inline tests/Tests/Api/PractitionerApiTest.php:12]: Guessing the above 2 lines can be removed

Comment (yashrajbothra) [reply tests/Tests/Api/PractitionerApiTest.php:12]: Definitely! 11 LN can be removed can't guess the second one?

Comment (bradymiller) [reply tests/Tests/Api/PractitionerApiTest.php:12]: oops, just meant line 11 :)

**Comment by bradymiller** [comment]
hi @yashrajbothra , Nice work!
Just noted some minor issues in the review (also looks like FHIR_README.md will need updating). After issues are addressed and is testing well for you, then let me know and will then bring this into codebase.

**Comment by yashrajbothra** [comment]
I think we are done with this resource (Only if Travis show Green flag) :)

**Comment by bradymiller** [comment]
Odd, getting some sporadic travis fails. Likely not issue with code in this PR but on travis end. Re-running tests now.

File: _rest_routes.inc.php [hunk 6]
```diff
@@ -387,6 +398,24 @@
...
+        RestConfig::authorization_check("patients", "demo");
+        return (new FhirPractitionerRestController())->getOne($id);
+    },
+    "POST /fhir/Practitioner" => function () {
+        RestConfig::authorization_check("patients", "demo");
+        $data = (array) (json_decode(file_get_contents("php://input"), true));
+        return (new FhirPractitionerRestController())->post($data);
+    },
+    "PATCH /fhir/Practitioner/:id" => function ($id) {
+        RestConfig::authorization_check("patients", "demo");
```
Comment (bradymiller) [inline _rest_routes.inc.php:415]: The above line in the 4 endpoints above should be this instead:
```
RestConfig::authorization_check("admin", "users");
```

**Comment by bradymiller** [comment]
@yashrajbothra , 
Looks like something is wrong with the sample data in one or more of the practitioner fhir fixtures, which is why it is failing intermittently (ie. dependent on which sample data it ends up using for the testing).

**Comment by bradymiller** [comment]
@yashrajbothra ,
Did some old fashioned debugging (outputted the fhir practitioner fixtures that were being used in the testing), and I think that "Harold Hippocrates" is the criminal wielding the `http://hl7.org.fhir/sid/us-npi` weapon :)

**Comment by bradymiller** [comment]
@yashrajbothra , Looks great and bringing this in! -brady

**Pull request merged**